File: python_sdk/src/reality_capture/service/detectors.py
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field
from datetime import datetime


class DetectorExport(Enum):
    OBJECTS = "Objects"
    LINES = "Lines"
    POLYGONS = "Polygons"
    LOCATIONS = "Locations"


class Capabilities(BaseModel):
    labels: list[str] = Field(description="Labels of the detector version.")
    exports: list[DetectorExport] = Field(description="Exports of the detector version.")


class DetectorStatus(Enum):
    AWAITING_DATA = "AwaitingData"
    READY = "Ready"


class DetectorVersion(BaseModel):
    creation_date: datetime = Field(description="Creation date of the version.", alias="creationDate")
    version: str = Field(description="Version number.")
    status: DetectorStatus = Field(description="Status of the version.")
    download_url: Optional[str] = Field(None, description="URL to download the detector version. "
                                                          "It is present only if the version status is 'Ready'.",
                                        alias="downloadUrl")
    creator_id: Optional[str] = Field(None, description="User Id of the version creator.", alias="creatorId")
    capabilities: Capabilities = Field(description="Capabilities of the version.")


class DetectorType(Enum):
    PHOTO_OBJECT_DETECTOR = "PhotoObjectDetector"
    PHOTO_SEGMENTATION_DETECTOR = "PhotoSegmentationDetector"
    ORTHOPHOTO_SEGMENTATION_DETECTOR = "OrthophotoSegmentationDetector"
    POINT_CLOUD_SEGMENTATION_DETECTOR = "PointCloudSegmentationDetector"


class DetectorBase(BaseModel):
    name: str = Field(description="Name of the detector.")
    display_name: Optional[str] = Field(None, description="Display name of the detector.", alias="displayName")
    description: Optional[str] = Field(None, description="Description of the detector.")
    type: DetectorType = Field(description="Type of the detector.")
    documentation_url: Optional[str] = Field(None, description="Display name of the detector.",
                                             alias="documentationUrl")


class Detector(DetectorBase):
    versions: list[DetectorVersion] = Field(description="All existing versions of the detector.")


class DetectorResponse(BaseModel):
    detector: Detector = Field(description="Detector.")


class DetectorMinimal(DetectorBase):
    latest_version: Optional[str] = Field(None, description="The latest version of the detector "
                                                            "with 'Ready' status, if any.",
                                          alias="latestVersion")


class DetectorsMinimalResponse(BaseModel):
    detectors: list[DetectorMinimal] = Field(description="List of minimal detectors.")

File: typescript_sdk/reality_capture/src/service/detectors.ts
/*import { z } from "zod";

export enum DetectorExport {
  OBJECTS = "Objects",
  LINES = "Lines",
  POLYGONS = "Polygons",
  LOCATIONS = "Locations",
}

export enum DetectorStatus {
  AWAITING_DATA = "AwaitingData",
  READY = "Ready",
}

export enum DetectorType {
  PHOTO_OBJECT_DETECTOR = "PhotoObjectDetector",
  PHOTO_SEGMENTATION_DETECTOR = "PhotoSegmentationDetector",
  ORTHOPHOTO_SEGMENTATION_DETECTOR = "OrthophotoSegmentationDetector",
  POINT_CLOUD_SEGMENTATION_DETECTOR = "PointCloudSegmentationDetector",
}

export const CapabilitiesSchema = z.object({
  labels: z.array(z.string()).describe("Labels of the detector version."),
  exports: z.array(z.nativeEnum(DetectorExport)).describe("Exports of the detector version."),
});
export type Capabilities = z.infer<typeof CapabilitiesSchema>;

export const DetectorVersionSchema = z.object({
  creationDate: z.coerce.date().describe("Creation date of the version."),
  version: z.string().describe("Version number."),
  status: z.nativeEnum(DetectorStatus).describe("Status of the version."),
  downloadUrl: z.string().optional().describe("URL to download the detector version. It is present only if the version status is 'Ready'."),
  creatorId: z.string().optional().describe("User Id of the version creator."),
  capabilities: CapabilitiesSchema.describe("Capabilities of the version."),
});
export type DetectorVersion = z.infer<typeof DetectorVersionSchema>;

export const DetectorBaseSchema = z.object({
  name: z.string().describe("Name of the detector."),
  displayName: z.string().optional().describe("Display name of the detector."),
  description: z.string().optional().describe("Description of the detector."),
  type: z.nativeEnum(DetectorType).describe("Type of the detector."),
  documentationUrl: z.string().optional().describe("Display name of the detector."),
});
export type DetectorBase = z.infer<typeof DetectorBaseSchema>;

export const DetectorSchema = DetectorBaseSchema.extend({
  versions: z.array(DetectorVersionSchema).describe("All existing versions of the detector."),
});
export type Detector = z.infer<typeof DetectorSchema>;

export const DetectorResponseSchema = z.object({
  detector: DetectorSchema.describe("Detector."),
});
export type DetectorResponse = z.infer<typeof DetectorResponseSchema>;

export const DetectorMinimalSchema = DetectorBaseSchema.extend({
  latestVersion: z.string().optional().describe("The latest version of the detector with 'Ready' status, if any."),
});
export type DetectorMinimal = z.infer<typeof DetectorMinimalSchema>;

export const DetectorsMinimalResponseSchema = z.object({
  detectors: z.array(DetectorMinimalSchema).describe("List of minimal detectors."),
});
export type DetectorsMinimalResponse = z.infer<typeof DetectorsMinimalResponseSchema>;*/